scripts/igmp_query_injector.py

```diff
@@ -1,232 +1,95 @@
+    Parameters:
+      self    The instantiation of an IGMP class
+      p       The IGMP message in hex in network byte order
+      pay     Additional payload for the IGMP message
     """
+    p += pay
+    if self.chksum is None:
+      ck = checksum(p)
+      p = p[:2]+chr(ck>>8)+chr(ck&0xff)+p[4:]
+    return p
 
-    xs_handler = xs()
-
-    @classmethod
-    def close(cls):
-        cls.xs_handler.close()
-
-    @classmethod
-    def read(cls, path):
-        return cls.xs_handler.read('', path)
-
-    @classmethod
-    def watch(cls, path, token):
-        return cls.xs_handler.watch(path, token)
+bind_layers(IP, IGMP, frag=0, proto=2)
+#################################
 
-    @classmethod
-    def unwatch(cls, path, token):
-        return cls.xs_handler.unwatch(path, token)
+# global xenstore handler
+g_xs_handler = xs()
 
-    @classmethod
-    def read_watch(cls):
-        return cls.xs_handler.read_watch()
 
-    @staticmethod
-    def vif_mac_path(domid, vifid):
-        return '/local/domain/%d/device/vif/%d/mac' % (domid, vifid)
-
-    @staticmethod
-    def vif_state_path(domid, vifid):
-        return '/local/domain/%d/device/vif/%d/state' % (domid, vifid)
+class ERRNO(object):
+    SUCC = 0
 
 
-class XSWatch(object):
-    __metaclass__ = ABCMeta
+class XSWatcher(object):
     """Tool for watching xenstore
     """
     def __init__(self):
-        self.watches = []
+        self.watches = dict()
 
     def watch(self, path, token):
-        self.watches.append(path)
-        return XSUtil.watch(path, token)
+        self.watches[path] = token
+        return g_xs_handler.watch(path, token)
 
     def unwatch(self, path, token):
-        self.watches.remove(path)
-        return XSUtil.unwatch(path, token)
-
-    def process_all_watches(self):
-        while self.watches:
-            path, token = XSUtil.read_watch()
-            if self.val_change_to_expected(path, token):
-                self.unwatch(path, token)
-                self.put_task(path, token)
-
-    @abstractmethod
-    def val_change_to_expected(self, path, token):
-        pass
-
-    @abstractmethod
-    def put_task(self, path, token):
-        pass
-
-
-class XSVifStateWatch(XSWatch):
-    def __init__(self, expected_state):
-        super(XSVifStateWatch, self).__init__()
-        self.expected_state = expected_state
-        self.queue = Queue.Queue()
+        self.watches.pop(path)
+        return g_xs_handler.unwatch(path, token)
 
-    def val_change_to_expected(self, path, vif):
-        return XSUtil.read(vif.state_path) == self.expected_state
-
-    def put_task(self, path, vif):
-        self.queue.put(InjectTask(vif, len(self.watches) == 0))
-
-    def get_task(self, timeout):
-        return self.queue.get(timeout=timeout)
-
-
-class InjectTask(object):
-    def __init__(self, vif, is_last):
-        self.vif = vif
-        self.is_last = is_last
+    def read_watch(self):
+        return g_xs_handler.read_watch()
 
 
 class Vif(object):
```

Comment:
I don't think there is much value in keeping this class.

Reply:
Accept